### PR TITLE
Add/Remove No Execute Taint

### PR DESCRIPTION
**Pull request by mshitrit**
- Adding a  `No Execute` taint in order to make sure after reboot the node is not running any workload - this is important so we can move this workload elsewhere even if the node does not have connectivity to the control plain.
- Once a node is healthy the taint is removed.

### CODE REVIEW

**Comment by beekhof** [comment]
Have you thought about putting it just before the snr.Spec.RemediationStrategy switch block?

**Review by slintes — CHANGES_REQUESTED** [review]
- seems there is some best practice about naming taints, e.g. `node.kubernetes.io/out-of-service=nodeshutdown:NoExecute`. Maybe use `medik8s.io/remediation=nodefencing:NoExecute`
- can we have a e2e test for the taint please?

**Comment by mshitrit** [comment]
> Maybe use `medik8s.io/remediation=nodefencing:NoExecute`

Sure :+1: 

> can we have a e2e test for the taint please?

Even though there is no specific e2e test for this use case my initial bug (not excluding the agent from the taint) was caught by our current e2e test so I'm not sure if adding another E2E is the way to go (I'd like to keep e2e tests to the necessary minimum).

In any case I'm planing to try to add a UT for this use case.
WDYT ?

**Comment by slintes** [comment]
> > can we have a e2e test for the taint please?
> 
> Even though there is no specific e2e test for this use case my initial bug (not excluding the agent from the taint) was caught by our current e2e test

Sorry, but this is a bad argumeny IMHO. Current e2e catching this bug doesn't mean it catches other bugs, e.g. not setting the taint at all.   

> so I'm not sure if adding another E2E is the way to go (I'd like to keep e2e tests to the necessary minimum).
> 

IMHO this taint qualifies for "minimum". But if you think a unit test is good enough, I'm fine with it.

File: controllers/controller_test.go [hunk 1]
```diff
@@ -354,6 +364,32 @@ func verifyFinalizerExists() {
...
+	By("Verify that node does not have NoExecute taint")
+	Eventually(isNoExecuteTaintExist(), 10*time.Second, 200*time.Millisecond).Should(BeFalse())
+}
+
+func verifyNoExecuteTaintExist() {
+	By("Verify that node has NoExecute taint")
+	Eventually(isNoExecuteTaintExist(), 10*time.Second, 200*time.Millisecond).Should(BeTrue())
+}
+
+func isNoExecuteTaintExist() func() bool {
```
Comment (slintes) [inline controllers/controller_test.go:377]: this can be made easier: no need to return a new func(), as this already is a func. You just need to pass in this func just by name without () in the Eventually above:

```
... Eventually(isNoExecuteTaintExist,...

func isNoExecuteTaintExist() (bool, error) { // no new func here...}
```

**Review by slintes — CHANGES_REQUESTED** [review]
just 2 comments about the test, otherwise lgtm

@beekhof do you agree with the taint name?

File: controllers/controller_test.go (hunk 1 above)
Comment (beekhof) [reply controllers/controller_test.go:377]: I might be inclined to have the value be the operator name, eg. `self-node-remediation`

Comment (slintes) [reply controllers/controller_test.go:377]: so `medik8s.io/remediation=self-node-remediation:NoExecute` ?
Works for me 👍🏼
